File: R/omniR/AdvOp/ls_frame.r
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function is intended to list the object names or the list[name=object] if verbose, in the provided frame or all frames along  #
#   | the call stack, by matching a specific pattern to the names and the predicate upon the objects                                    #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |Scenarios:                                                                                                                         #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |[1] Search for certain pattern of functions within current session                                                                 #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |frame       :   <frame> object in which to search for objects                                                                      #
#   |                [NULL        ] <Default> Search in all frames along the call stack                                                 #
#   |                [frame       ]           Dedicated <frame> in which to search the objects                                          #
#   |predicate   :   Function predicate to apply to the objects as found, only those with True predicates will be returned              #
#   |                [<see def.>  ] <Default> Do not apply predicate                                                                    #
#   |                [function    ]           Function with the first argument to be applied upon the object as found, and return bool  #
#   |pattern     :   Regex pattern to search within the object names, used for <str_detect> instead of <str_match>                      #
#   |                [<see def.>  ] <Default> Search for all names without filtration                                                   #
#   |                [str         ]           Valid Regex string representation                                                         #
#   |verbose     :   Whether to return verbose results                                                                                  #
#   |                [FALSE       ] <Default> Only return a list of names found by the conditions                                       #
#   |                [TRUE        ]           Return a dict[name:object]                                                                #
#   |...         :   Options for <stringr::regex> to compile a valid Regex parser                                                       #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |<Various>   :   This function output different values in below convention:                                                         #
#   |                [1] If <verbose == FALSE>, return a [character vector] of names matching the conditions                            #
#   |                [2] If <verbose == TRUE>, return a [named list] of names pairing the objects, which match the conditions           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20240219        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Version 1.                                                                                                                  #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent Modules                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |magrittr, stringr, stringi, rlang                                                                                              #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent user-defined functions                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------------#

#001. Append the list of required packages to the global environment
#Below expression is used for easy copy-paste from raw text strings instead of quoted ones.
lst_pkg <- deparse(substitute(c(
	magrittr, stringr, stringi, rlang
)))
#Quote: https://www.regular-expressions.info/posixbrackets.html?wlr=1
lst_pkg <- paste0(lst_pkg, collapse = '')
lst_pkg <- gsub('[[:space:]]', '', lst_pkg, perl = T)
lst_pkg <- gsub('^c\\((.+)\\)', '\\1', lst_pkg, perl = T)
lst_pkg <- unlist(strsplit(lst_pkg, ',', perl = T))
options( omniR.req.pkg = base::union(getOption('omniR.req.pkg'), lst_pkg) )

#We should use the pipe operands supported by below package
library(magrittr)

ls_frame <- function(
	frame = NULL
	,predicate = function(x){TRUE}
	,pattern = '.*'
	,verbose = FALSE
	,...
){
	#001. Handle parameters
	#[Quote: https://stackoverflow.com/questions/15595478/how-to-get-the-name-of-the-calling-function-inside-the-called-routine ]
	LfuncName <- deparse(sys.call()[[1]])
	#If above statement cannot find the name correctly, this function must have been called via [do.call] or else,
	# hence we need to traverse one layer above current one and extract the first argument of that call.
	if (grepl('^function.+$',LfuncName[[1]],perl = T)) LfuncName <- gsub('^.+?\\((.+?),.+$','\\1',deparse(sys.call(-1)),perl = T)[[1]]
	if (!is.function(predicate)) predicate <- function(x){TRUE}
	if (!is.logical(verbose)) verbose <- FALSE

	#050. Local parameters
	kw <- rlang::list2(...)

	#100. Setup Regex
	#110. Retrieve the keyword arguments
	params_r <- formalArgs(stringr::regex)
	params_i_coll <- formalArgs(stringi::stri_opts_collator)
	params_i_rx <- formalArgs(stringi::stri_opts_regex)
	params_i_br <- formalArgs(stringi::stri_opts_brkiter)

	#130. Obtain all defaults of keyword arguments of the function
	kw_raw <- c(
		params_r[!params_r %in% c('pattern')]
		,params_i_coll
		,params_i_rx
		,params_i_br
	) %>%
		{.[!. %in% c('...')]}

	#150. Create the final keyword arguments for calling the function
	kw_final <- kw[names(kw) %in% kw_raw]

	#190. Parse the pattern
	ptn <- do.call(stringr::regex, c(list(pattern = pattern), kw_final))

	#500. Define helper functions
	#[Quote: https://stackoverflow.com/questions/11885207/get-all-parameters-as-list ]
	h_globframe <- function(frame) {
		evalq(as.list(environment()), envir = frame) %>%
			{Filter(predicate, .)} %>%
			{.[stringr::str_detect(names(.), ptn)]}
	}

	#600. Directly export when a frame is specified
	if (is.environment(frame)) {
		rstOut <- h_globframe(frame)
		if (verbose) {
			return(rstOut)
		} else {
			return(names(rstOut))
		}
	}

	#700. Search starting from the parent frame and backwards
	ifr <- 1
	rstOut <- list()
	while (T) {
		#100. Retrieve the content of the parent frame to the previous one
		pframe <- parent.frame(ifr)

		#500. Glob the frame
		rstInt <- h_globframe(pframe)

		#700. Merge the results by taking the objects located in the child frames as higher priority
		rstOut <- c(
			rstOut
			,rstInt[!names(rstInt) %in% names(rstOut)]
		)

		#800. Stop the loop if current frame is the global environment
		#Quote: https://www.r-bloggers.com/2011/06/environments-in-r/
		if (environmentName(pframe) == 'R_GlobalEnv') break

		#900. Increment the counter of parent frames
		ifr <- ifr + 1
	}

	#999. Export
	if (verbose) {
		return(rstOut)
	} else {
		return(names(rstOut))
	}
}

#[Full Test Program;]
if (FALSE){
	#Simple test
	if (TRUE){
		#100. Create objects
		aa1 <- 1
		aa2 <- function(){}
		aa3 <- list()
		ab1 <- 'aa'

		#300. Simple test
		ls_frame(pattern = '^aa')
		# [1] "aa1" "aa2" "aa3"

		ls_frame(pattern = '^aa', verbose = TRUE)
		# $aa1
		# [1] 1
		#
		# $aa2
		# function(){}
		#
		# $aa3
		# list()

		ls_frame(pattern = '^aa', predicate = is.list)
		# [1] "aa3"

		#400. Test the search within specific scopes
		testscope <- function(){
			frame <- environment()

			aa2 <- function(){'local'}
			aa3 <- list('1' = 3)

			#100. Include all global variables, with those local variables found prior to the globals if they share the same names
			print('All:')
			print(ls_frame(pattern = '^aa'))

			#300. Only include local variables defined in current frame
			print('Locals:')
			print(ls_frame(frame = frame, pattern = '^aa'))
		}
		testscope()
		# [1] "All:"
		# [1] "aa3" "aa2" "aa1"
		# [1] "Locals:"
		# [1] "aa3" "aa2"

		#500. Verify whether the members of <ls_frame> can be obtained
		ls_frame(pattern = 'h_globframe')
		# NULL

		ls_frame(pattern = 'h_globframe', verbose = T)
		# list()

	}
}

File: Python/omniPy/AdvOp/ls_frame.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys, re
import inspect
from collections.abc import Callable
from typing import Any

def ls_frame(
    frame = None
    ,predicate : Callable[[Any, ...], bool] = lambda x: True
    ,scope : str | list[str] = ['f_locals','f_globals']
    ,pattern : str | re.Pattern = r'.*'
    ,flags : int = re.NOFLAG
    ,verbose : bool = False
) -> list[str] | dict[str, Any]:
    #000. Info.
    '''
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function is intended to list the object names or the dict[name:object] if verbose, in the provided frame or all frames along  #
#   | the call stack, by matching a specific pattern to the names and the predicate upon the objects                                    #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |Scenarios:                                                                                                                         #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |[1] Search for certain pattern of functions within current session                                                                 #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |frame       :   <frame> object in which to search for objects                                                                      #
#   |                [None        ] <Default> Search in all frames along the call stack                                                 #
#   |                [frame       ]           Dedicated <frame> in which to search the objects                                          #
#   |predicate   :   Callable predicate to apply to the objects as found, only those with True predicates will be returned              #
#   |                [<see def.>  ] <Default> Do not apply predicate                                                                    #
#   |                [callable    ]           Callable with the first argument to be applied upon the object as found, and return bool  #
#   |scope       :   Which scope to search for the variables in the frames along the call stacks                                        #
#   |                IMPORTANT: This must be provided a sequence or a single string, indicating the search order of the scopes          #
#   |                [<see def.>  ] <Default> Search for the variables in <f_locals> and then <f_globals> until the last try            #
#   |                [f_locals    ]           Only search for the variables in <f_locals> of every frame along the call stacks          #
#   |                [f_globals   ]           Only search for the variables in <f_globals> of every frame along the call stacks         #
#   |pattern     :   Regex pattern to search within the object names, used for <re.Search> instead of <re.Match>                        #
#   |                [<see def.>  ] <Default> Search for all names without filtration                                                   #
#   |                [str         ]           Valid Regex string representation                                                         #
#   |                [re.Pattern  ]           Valid Pattern object compiled by <re.compile>                                             #
#   |flags       :   Regex flags to create the internal pattern if <pattern> is provided a string                                       #
#   |                [<see def.>  ] <Default> No certain flag is applied                                                                #
#   |                [flags       ]           Valid flags defined in package <re>                                                       #
#   |verbose     :   Whether to return verbose results                                                                                  #
#   |                [False       ] <Default> Only return a list of names found by the conditions                                       #
#   |                [True        ]           Return a dict[name:object]                                                                #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |<Various>   :   This function output different values in below convention:                                                         #
#   |                [1] If <verbose == False>, return a [list] of names matching the conditions                                        #
#   |                [2] If <verbose == True>, return a [dict] of names pairing the objects, which match the conditions                 #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20240218        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Version 1.                                                                                                                  #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent Modules                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |sys, re, inspect, collections, typing                                                                                          #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent user-defined functions                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------------#
    '''

    #010. Check parameters.
    #011. Prepare log text.
    #python 动态获取当前运行的类名和函数名的方法: https://www.cnblogs.com/paranoia/p/6196859.html
    LfuncName : str = sys._getframe().f_code.co_name

    #012. Parameter buffer
    vld_scope = ['f_locals','f_globals']
    if isinstance(scope, str):
        scope = [scope]
    vfy_scope = [ s for s in scope if s not in vld_scope ]
    if len(vfy_scope):
        raise ValueError(f'[{LfuncName}][scope] must be among {str(vld_scope)}!')
    if isinstance(pattern, str):
        ptn = re.compile(pattern, flags)
    elif isinstance(pattern, re.Pattern):
        ptn = pattern
    else:
        raise ValueError(f'[{LfuncName}][pattern] cannot be compiled as Regex, given {str(pattern)}!')

    #050. Local parameters

    #100. Define helper functions
    #[ASSUMPTION]
    #[1] When searching for a variable in both scopes, while it exists in both, the search result is only from <f_globals>
    #     of current frame <Python <= 3.11.2>
    def h_globframe(frame):
        rstOut = {}
        for s in scope:
            rstOut = {
                **rstOut
                ,**{
                    k:v
                    for k,v in frame.__getattribute__(s).items()
                    if ptn.search(k) and predicate(v) and (k not in rstOut)
                }
            }
        return(rstOut)

    #300. Directly export when a frame is specified
    if inspect.isframe(frame):
        rstOut = h_globframe(frame)
        if verbose:
            return(rstOut)
        else:
            return(list(rstOut.keys()))

    #500. Search starting from the parent frame and backwards
    frame = sys._getframe(1)
    rstOut = {}
    #Avoid errors to be raised when reaching the global environment
    #Quote: https://stackoverflow.com/questions/39265823/python-sys-getframe
    while frame:
        rstOut = {
            **rstOut
            ,**{
                k:v
                for k,v in h_globframe(frame).items()
                if k not in rstOut
            }
        }
        frame = frame.f_back

    #700. Search within [global] environment
    rstOut = {
        **rstOut
        ,**{
            k:v
            for k,v in globals().items()
            if ptn.search(k) and predicate(v) and (k not in rstOut)
        }
    }

    #900. Output
    if verbose:
        return(rstOut)
    else:
        return(list(rstOut.keys()))
#End ls_frame

'''
#-Notes- -Begin-
#Full Test Program[1]:
if __name__=='__main__':
    #010. Create envionment.
    import sys
    from functools import partial
    dir_omniPy : str = r'D:\Python\ '.strip()
    if dir_omniPy not in sys.path:
        sys.path.append( dir_omniPy )
    from omniPy.AdvOp import ls_frame

    #100. Create objects
    aa1 = 1
    def aa2(): pass
    aa3 = {}
    ab1 = 'aa'

    #200. Define predicates for the function to filter objects
    def pr_instance(obj, instance):
        if instance is callable:
            return(callable(obj))
        else:
            return(isinstance(obj, instance))

    #300. Simple test
    ls_frame(pattern = r'^aa')
    # ['aa1', 'aa2', 'aa3']

    ls_frame(pattern = r'^aa', verbose = True)
    # {'aa1': 1, 'aa2': <function __main__.aa2()>, 'aa3': {}}

    ls_frame(pattern = r'^aa', predicate = partial(pr_instance, instance = dict))
    # ['aa3']

    #400. Test the search within specific scopes
    def testscope():
        frame = sys._getframe()

        def aa2(): pass
        aa3 = {1 : 3}

        #100. Include all global variables, with those local variables found prior to the globals if they share the same names
        print('All:')
        print(ls_frame(frame = frame, pattern = r'^aa'))

        #300. Only include local variables defined in current frame
        print('Locals:')
        print(ls_frame(frame = frame, pattern = r'^aa', scope = 'f_locals'))

    testscope()
    # All:
    # ['aa2', 'aa3', 'aa1']
    # Locals:
    # ['aa2', 'aa3']

    #500. Verify whether the members of <ls_frame> can be obtained
    ls_frame(pattern = r'h_globframe')
    # []
#-Notes- -End-
'''
